docs/manual/generics.tex

```diff
@@ -850,6 +854,174 @@
   void m(@Nullable Object obj)
 \end{Verbatim}
 
+\sectionAndLabel{Class qualifier parameters}{class-qualifier-polymorphism}
+
+The Checker Framework supports type \emph{qualifier} polymorphism for classes.
+Java generic classes permit you to supply a type argument to a generic
+class, as in \code{List<@Nullable String>}.
+Class qualifier parameters permit you to supply a type qualifier (only,
+without a Java basetype) to any
+class (generic or not).
+This is useful when the class hard-codes a type internally (as
+\<StringBuffer> hard-codes \<Character>), but you wish to apply a qualifier
+to it, such as to distinguish \<StringBuffer>s that contain tainted
+characters from those that contain untainted characters.  (For a discussion
+of tainting, see \chapterpageref{tainting-checker}.)
+
+To add a qualifier parameter to a class, annotate its declaration with \refqualclass{framework/qual}{HasQualifierParameter}
+and write the class of the top qualifier as its element.
+
+\begin{Verbatim}
+@HasQualifierParameter(Tainted.class)
+class StringBuffer { ... }
+\end{Verbatim}
+
+A qualifier on a use of \<StringBuffer> is treated as
+appearing both on the \<StringBuffer> and on its conceptual type argument.  That
+is:\\
+\begin{tt}
+  @Tainted StringBuffer $\approx$ @Tainted Collection<@Tainted Character>\\
+  @Untainted StringBuffer $\approx$ @Untainted Collection<@Untainted Character>
+\end{tt}
+
+If two types have different qualifier arguments, they have no subtyping
+relationship.  (This is ``invariant subtyping'', also used by Java for
+generic classes.)  In particular, \<@Untainted StringBuffer> is not a
+subtype of \<@Tainted StringBuffer>; an attempt to cast between them, in
+either direction, will yield an \<invariant.cast.unsafe> error.
```

Comment:
I don't understand this question.  Can you restate it?

Reply:
This states that `@Untainted StringBuffer` and `@TaintedStringBuffer` are not comparable.
How do I declare a field that can take either kind of `StringBuffer`, without also adding a qualifier parameter to that class?
How does local variable type refinement work? We usually start with the top type and then refine to subtypes. Now that the two possible types are incomparable, how are local variables handled?

Reply:
>How do I declare a field that can take either kind of StringBuffer, without also adding a qualifier parameter to that class?

You can't.

>How does local variable type refinement work? We usually start with the top type and then refine to subtypes. Now that the two possible types are incomparable, how are local variables handled?

If the type of a local variable has a qualifier parameter, it can't be refined.

Reply:
Mentioning these in the manual would be useful to understand how to use this feature.